### PR TITLE
(PIE-281) Revive Splunk Acceptance Tests

### PR DESCRIPTION
**Pull request by gsparks**
This PR migrates much of the acceptance testing infrastructure from the
reporting module such that now a puppet master node is created and
installed on vmpooler with the splunk container running within it.

No new acceptance tests have been added. but the existing tests run once
again.

Additionally the splunk container is started and ready to accept HEC
requests for further acceptance testing. The Puppet Report Viewer
Splunkbase app is installed into the Splunk container so that we can now
send HEC requests to the container and verify that they were indexed
correctly.

A jenkins project has been created.
https://cinext-jenkinsmaster-pipeline-prod-1.delivery.puppetlabs.net/view/PIE%20Team/job/pipeline_puppetlabs-splunk_hec/

### CODE REVIEW

**Comment by gsparks** [comment]
I'm not sure how we feel about having both Github actions and Jenkins running simultaneously, but I figured I'd leave it until we have consensus. Right now Actions handles the static checks / deploy process and jenkins handles the acceptance tests.

**Comment by gsparks** [comment]
Also I figured I'd add additional acceptance tests later as this PR is large enough as it is.

File: Rakefile [hunk 1]
```diff
@@ -75,3 +75,123 @@ EOM
...
+    }]
+    write_to_inventory_file(inventory_hash, 'inventory.yaml')
+  end
+
+  desc 'Installs the module on the master'
+  task :install_module do
+    Rake::Task['litmus:install_module'].invoke(master.uri)
+  end
+
+  # splunk_hec.yaml must be deleted from the master if this is run more than once.
```
Comment (ekinanp) [inline Rakefile:143]: I'd go ahead and do this b/c I see no good reason to have the user do this.. Maybe `master.run_shell('rm -rf <splunk_hec.yaml>')`.

File: spec/acceptance/class_spec.rb [hunk 2]
```diff
@@ -24,7 +24,7 @@ class { 'splunk_hec': url => "http://httpbin.org/post", token => '', record_even
     end
 
     it 'Successfully sends data to an http endpoint' do
-      run_shell('cat /etc/puppetlabs/code/environments/production/modules/splunk_hec/examples/orchestrator_metrics.json | puppet splunk_hec --sourcetype puppet:summary --saved_report')
+      run_shell('cat /etc/puppetlabs/code/environments/production/modules/splunk_hec/examples/orchestrator_metrics.json | puppet splunk_hec --sourcetype puppet:summary')
```
Comment (ekinanp) [inline spec/acceptance/class_spec.rb:27]: Why this change?

File: spec/support/acceptance/configure_splunk_hec.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Comment (ekinanp) [inline spec/support/acceptance/configure_splunk_hec.sh:1]: Why is this necessary? Shouldn't testing the setup be included in the acceptance tests?

File: spec/support/acceptance/splunk/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+version: "2.0"
+
+services:
+  enterprise:
+    image: splunk/splunk:latest
+    hostname: splunk_instance
+    environment:
+      - SPLUNK_START_ARGS=--accept-license
+      - ASFDGASDFG=5
```
Comment (ekinanp) [inline spec/support/acceptance/splunk/docker-compose.yml:9]: What is this variable for?

File: spec/support/acceptance/splunk/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+      # The splunkbase credentials are to download and install the Puppet
+      # Report Viewer from Splunkbase.
+      # We could alternatively download the packaged app from a location
+      # like github for testing.
+      - SPLUNK_APPS_URL=https://splunkbase.splunk.com/app/4413/release/3.0.1/download
```
Comment (ekinanp) [inline spec/support/acceptance/splunk/docker-compose.yml:14]: This is a persistent instance right?

File: spec/support/acceptance/splunk/default.yml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+---
```
Comment (ekinanp) [inline spec/support/acceptance/splunk/default.yml:1]: What's this file for?

File: Gemfile [hunk 7]
```diff
@@ -25,15 +25,16 @@ group :development do
...
   gem "puppet-module-posix-default-r#{minor_version}", '~> 0.3', require: false, platforms: [:ruby]
   gem "puppet-module-posix-dev-r#{minor_version}", '~> 0.3',     require: false, platforms: [:ruby]
+  gem "puppet_litmus", '~> 0.18',                                require: false, platforms: [:ruby]
   gem "puppet-module-win-default-r#{minor_version}", '~> 0.3',   require: false, platforms: [:mswin, :mingw, :x64_mingw]
   gem "puppet-module-win-dev-r#{minor_version}", '~> 0.3',       require: false, platforms: [:mswin, :mingw, :x64_mingw]
   gem "github_changelog_generator",                              require: false, git: 'https://github.com/skywinder/github-changelog-generator', ref: '20ee04ba1234e9e83eb2ffb5056e23d641c7a018' if Gem::Version.new(RUBY_VERSION.dup) >= Gem::Version.new('2.2.2')
-end
-group :system_tests do
+  gem "rspec_junit_formatter",                                   require: false
+  gem "rspec-puppet", '= 2.7.10',                                require: false
```
Comment (RandomNoun7) [inline Gemfile:33]: I think this should be slightly more loosely pinned. Something like `>= 2.7.10 < 3` or whatever syntax like that works.

File: spec/support/acceptance/splunk/docker-compose.yml (hunk 4 above)
Comment (gsparks) [reply spec/support/acceptance/splunk/docker-compose.yml:9]: nothing, thanks! Will remove.

File: spec/support/acceptance/splunk/docker-compose.yml (hunk 5 above)
Comment (gsparks) [reply spec/support/acceptance/splunk/docker-compose.yml:14]: Yeah this is just a splunkbase account that is used to download splunk apps. It's separate from a login to a splunk instance.

File: spec/spec_helper_acceptance_local.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
+# rubocop:disable Style/AccessorMethodName
+
+require 'serverspec'
+require 'puppet_litmus'
+require './spec/support/acceptance/helpers.rb'
```
Comment (RandomNoun7) [inline spec/spec_helper_acceptance_local.rb:5]: I don't think you need the `'./'` part of the path. Paths under `spec` are automatically added to `$LOAD_PATH` when running rspec.

File: spec/support/acceptance/splunk/default.yml (hunk 6 above)
Comment (gsparks) [reply spec/support/acceptance/splunk/default.yml:1]: This file is passed into the container on startup and is used to configure all the splunk settings you'd normally configure via the web interface.

[https://splunk.github.io/docker-splunk/ADVANCED.html](https://splunk.github.io/docker-splunk/ADVANCED.html)

File: spec/spec_helper_acceptance_local.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    # Stop the puppet service on the master to avoid edge-case conflicting
+    # Puppet runs (one triggered by service vs one we trigger)
+    master.run_shell('puppet resource service puppet ensure=stopped')
+  end
+end
+
+# TODO: This will cause some problems if we run the tests
+# in parallel. For example, what happens if two targets
+# try to modify site.pp at the same time?
+def set_sitepp_content(manifest)
```
Comment (RandomNoun7) [inline spec/spec_helper_acceptance_local.rb:22]: Would be great to get this as a helper into Litmus after Enis' PR is merged. Just a thought.

File: spec/support/acceptance/splunk/default.yml (hunk 6 above)
Comment (RandomNoun7) [reply spec/support/acceptance/splunk/default.yml:1]: Isn't this for configuring the splunk container?

File: spec/support/acceptance/splunk/default.yml [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  password: "{{ splunk_password | default(<password>) }}"
+  svc_port: 8089
+  s2s_port: 9997
+  http_port: 8000
+  hec:
+    enable: True
+    ssl: False
+    port: 8088
+    token: abcd1234
+  smartstore: null
```
Comment (RandomNoun7) [inline spec/support/acceptance/splunk/default.yml:19]: nit: missing newline.

File: spec/support/acceptance/splunk/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+      - SPLUNK_PASSWORD=piepiepie
+    volumes:
+      # default.yml is a mechanism to load splunk settings that would normally
+      # be configured through the ui.
+      - ./default.yml:/tmp/defaults/default.yml
+    ports:
+      # localhost:8000 will bring up the web interface
+      - "127.0.0.1:8000:8000"
+      # 8088 is the hec endpoint
+      - "127.0.0.1:8088:8088"
```
Comment (RandomNoun7) [inline spec/support/acceptance/splunk/docker-compose.yml:27]: Changing these bindings to `0.0.0.0`  might make things easier later for things like troubleshooting the tests and prevent an annoying bug for someone later on.

File: spec/support/acceptance/splunk_hec.yaml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+---
+"url" : "http://localhost:8088/services/collector/event"
+"token" : "abcd1234"
+"enable_reports" : "true"
+"manage_routes" : "true"
```
Comment (RandomNoun7) [inline spec/support/acceptance/splunk_hec.yaml:5]: nit: Newline

File: spec/support/acceptance/splunk/default.yml (hunk 6 above)
Comment (ekinanp) [reply spec/support/acceptance/splunk/default.yml:1]: Ah ok. I'd maybe post a comment in the file along those lines. May be nice to link to a reference? E.g. something like https://splunk.github.io/docker-splunk/ADVANCED.html

File: spec/support/acceptance/configure_splunk_hec.sh (hunk 3 above)
Comment (gsparks) [reply spec/support/acceptance/configure_splunk_hec.sh:1]: Turns out this isn't neccesary. Thanks!

File: spec/support/acceptance/splunk/docker-compose.yml (hunk 5 above)
Comment (gsparks) [reply spec/support/acceptance/splunk/docker-compose.yml:14]: I tried to create a team account, but team-pie@puppet.com already has an account. I'll update it as soon as I find the password haha.

File: spec/spec_helper_acceptance_local.rb (hunk 9 above)
Comment (gsparks) [reply spec/spec_helper_acceptance_local.rb:22]: Will plan to update once we get this into Litmus.

File: spec/spec_helper_acceptance_local.rb (hunk 8 above)
Comment (gsparks) [reply spec/spec_helper_acceptance_local.rb:5]: hmm, I'm getting a load error when I remove the `./`.

**Comment by RandomNoun7** [comment]
You can try just support/acceptance/blah, and if that’s still broken then
disregard completely because it’s not worth putting a lot of effort into
troubleshooting.

File: Rakefile (hunk 1 above)
Comment (gsparks) [reply Rakefile:143]: adding this to the `install_module` task, and removing the `configure_module` task.

File: spec/support/acceptance/splunk/default.yml (hunk 6 above)
Comment (gsparks) [reply spec/support/acceptance/splunk/default.yml:1]: will do!

File: spec/spec_helper_acceptance_local.rb (hunk 8 above)
Comment (gsparks) [reply spec/spec_helper_acceptance_local.rb:5]: `support/acceptance/helpers.rb` seems to be working.

File: spec/acceptance/class_spec.rb (hunk 2 above)
Comment (gsparks) [reply spec/acceptance/class_spec.rb:27]: The `--saved_report` option will submit the request immediately to the splunk instance configured in splunk_hec.yaml. It will not work currently as that file is not yet fully configured with the credentials for the splunk container we are running. I'm planning a PR right after this one to add this back and more acceptance tests that actually communicate with the Splunk container. I could also add it for just this test in this PR as well if we feel that changing this test feels janky. Happy to go either way.

Comment (ekinanp) [reply spec/acceptance/class_spec.rb:27]: Ah ok, that's fine. Was just curious.

**Comment by gsparks** [comment]
Fixed an issue where the `--saved_report `argument for the report processor caused the report to fail because the report data did not include a sourcetype which is used to specify different HEC tokens for different puppet report types eg.(summary, facts). Sourcetype should not always be required for this kind of testing purpose.

It looks like this test probably never ran correctly given the sample json report data that is in the repo.

Added a readme snippet about running the tests.

File: lib/puppet/util/splunk_hec.rb [hunk 13]
```diff
@@ -45,10 +45,17 @@ def create_http(source_type)
...
-    token_name = "token_#{source_type}"
-    http = create_http(source_type)
-    token = settings[token_name] || settings['token'] || raise(Puppet::Error, 'Must provide token parameter to splunk class')
+    if body['sourcetype']
+      source_type = body['sourcetype'].split(':')[1]
+      token_name = "token_#{source_type}"
+      http = create_http(source_type)
+      token = settings[token_name] || settings['token'] || raise(Puppet::Error, 'Must provide token parameter to splunk class')
+    else
+      # For testing purposes there isn't always a sourcetype.
```
Comment (ekinanp) [inline lib/puppet/util/splunk_hec.rb:54]: Seems weird that this is only here to get the tests to pass. Is it an important precondition of this module that the `sourcetype` must exist? If so, may be worth finding a way to get that test to pass _without_ having this fix. That way can be done in a follow-up PR though.

File: spec/spec_helper_acceptance_local.rb (hunk 9 above)
Comment (ekinanp) [reply spec/spec_helper_acceptance_local.rb:22]: We don't need the TODO anymore btw, that was there back before the multi-node Litmus crap.

File: lib/puppet/util/splunk_hec.rb (hunk 13 above)
Comment (gsparks) [reply lib/puppet/util/splunk_hec.rb:54]: Splunk should be able to receive data without a specific sourcetype although the module isn't currently set up to do this automatically. It's nice to be able to do this for testing purposes, however an alternative approach would be to specify a sourcetype which does not exist, which creates the same behavior. That would be a way to keep this test, without modifying the production code.

Comment (ekinanp) [reply lib/puppet/util/splunk_hec.rb:54]: Yeah, let's do the "nonexistent sourcetype" option then if it's easy enough.

Comment (gsparks) [reply lib/puppet/util/splunk_hec.rb]: Cool, updated!
